### PR TITLE
Creating a C parser 

### PR DESCRIPTION
**Pull request by Sc0rpi0n101**
#### References to other Issues or PRs

#16563

#### Brief description of what is fixed or changed

This PR contains the initial implementation of the C parser for SymPy using Clang' s python API. 
This project is not yet complete and under construction.

#### TODO list

- [x] Make the module name all lowercase
- [x] Move the tests into a tests/ directory
- [x] Make sure the external dependencies are optional
- [x] Make sure the external dependencies are installed in Travis for the tests
- [x] Mark the module as provisional so that the API can be changed later if necessary
- [x] Add a release notes entry
- [x] Add the package to the conda-forge

#### Release Notes

* parsing
  * Added a new Parser to parse C source code.

### CODE REVIEW

File: sympy/parsing/C/parser.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    UnaryOperator
+)
+
+import argparse
+import os
+import re
+import sys
+
+
+class Writer(object):
```
Comment (czgdp1807) [inline sympy/parsing/C/parser.py]: Why no inheriting `Basic`? Just asking out of curiosity.

File: sympy/parsing/C/parser.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    def __init__(self, out,):
```
Comment (czgdp1807) [inline sympy/parsing/C/parser.py]: I think `,` shouldn't be there after `out`, right?

File: sympy/parsing/C/parser.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    def remove_big_block(self):
+        if not self.line_cleared:
+            self.out.write('\n')
+            self.line_cleared = True
+        while self.blank_lines < max(1, 2 - self.depth):
+            self.blank_lines += 1
+
+
+
+class BaseParser(object):
```
Comment (czgdp1807) [inline sympy/parsing/C/parser.py]: > Why not inheriting `Basic`? Just asking out of curiosity.

+1

File: sympy/parsing/C/parser.py (hunk 2 above)
Comment (Sc0rpi0n101) [reply sympy/parsing/C/parser.py]: Oh I left a comma 
I'll fix that

File: sympy/parsing/C/parser.py (hunk 1 above)
Comment (Sc0rpi0n101) [reply sympy/parsing/C/parser.py]: No particular reason.
I can inherit from basic if suggest so?

**Comment by certik** [comment]
Thanks @czgdp1807 for the review!

File: sympy/parsing/C/parser.py (hunk 1 above)
Comment (czgdp1807) [reply sympy/parsing/C/parser.py]: I think you should do it.

Comment (certik) [reply sympy/parsing/C/parser.py]: What would be the advantage of inheriting from Basic? I don't think this class is related to it at all.

Comment (czgdp1807) [reply sympy/parsing/C/parser.py]: I am suggesting for consistency.  If it's not required then no issues.  We can go without it.

Comment (certik) [reply sympy/parsing/C/parser.py]: I would only inherit from Basic if the class is supposed to be part of a SymPy expression, which this one does not.

Comment (czgdp1807) [reply sympy/parsing/C/parser.py]: Please resolve this conversation. Thanks @certik for the reason.

File: sympy/parsing/C/parser.py (hunk 3 above)
Comment (czgdp1807) [reply sympy/parsing/C/parser.py]: Please resolve this conversation.

**Comment by certik** [comment]
@NikhilPappu can you please fix the things that @czgdp1807 pointed out?

**Comment by certik** [comment]
@Sc0rpi0n101 Now you should add tests for this. Ideally they would be strings of C code in the Python test files, that would be passed to Clang and through your converter, and it would check that the correct SymPy expressions were obtained.

File: sympy/parsing/C/parser.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            converter.output('%s.py' % args.output, out)
+    else:
+        print("Not Implemented") #To be done
+
+    py_file = open('%s.py' % args.output, "r")
+    py_src = py_file.read()
+    py_file.close()
+    sym_src =  parse_expr(
+        py_src, transformations=(standard_transformations + (implicit_multiplication_application,)))
+    op_file = open('%s.py' % args.output, "w")
```
Comment (asmeurer) [inline sympy/parsing/C/parser.py]: Use `with open(...)`

File: sympy/parsing/C/parser.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+            return function
+
+     def transform_var_decl(self, node, context):
+        try:
+            children = node.get_children()
+            prev_child = None
+            child = next(children)
+            # print("FIRST CHILD", child.kind)
+
+def convert_file(src_in, src_out):
```
Comment (asmeurer) [inline sympy/parsing/C/parser.py]: If this is a command line tool, it should have 

```py
if __name__ == '__main__':
    main()
```

at the bottom so it can be called like `python -m sympy.parsing.c.parser`

File: .travis.yml [hunk 6]
```diff
@@ -272,6 +272,7 @@ before_install:
         export LD_LIBRARY_PATH=$CONDA_PREFIX/lib;
         export SYMPY_STRICT_COMPILER_CHECKS=1;
         conda clean --all;
+        pip install clang;
```
Comment (asmeurer) [inline .travis.yml]: This should be installed in the optional dependency build only.

Also, it should install with conda if possible. If this is missing from conda-forge, let's work on adding it.

File: setup.py [hunk 7]
```diff
@@ -429,5 +430,8 @@ def run(self):
           install_requires=[
             'mpmath>=%s' % mpmath_version,
             ],
+          extra_require={
+            'clang' : ["clang"],
```
Comment (asmeurer) [inline setup.py]: Currently we don't have any optional dependencies listed in setup.py. I would leave this out for now. If we want to add optional dependencies to setup.py, we should do so in a separate PR that adds all of them.

Comment (Sc0rpi0n101) [reply setup.py]: ok, I'll leave that out for now.

**Review by certik — APPROVED** [review]
I think it looks good to me now.

**Comment by certik** [comment]
I can't find the C parser to be executed at Travis. I would expect to see a line with `test_c_parser.py` at https://travis-ci.org/sympy/sympy/jobs/571465134, but it's not there.

**Comment by certik** [comment]
The Fortran parser is here: https://travis-ci.org/sympy/sympy/jobs/571465135, but the C parser is not tested.

**Review by certik — CHANGES_REQUESTED** [review]
We need to make sure the parser tests are actually executed.

**Comment by Sc0rpi0n101** [comment]
@certik The tests are being executed and passing now.

Are there any more changes needed?

**Review by certik — APPROVED** [review]

**Comment by certik** [comment]
Thank you, I think this looks good to merge.
